microscope/lasers/sapphire.py

```diff
@@ -0,0 +1,211 @@
+        elif status == '2':
+            return 'Warmup'
+        elif status == '3':
+            return 'Standby'
+        elif status == '4':
+            return 'Laser on'
+        elif status == '5':
+            return 'Laser ready'
+        elif status == '6':
+            return 'Error'
+        else:
+            return 'Undefined'
+
+    @lock_comms
+    def get_status(self):
+        result = []
+
+        result.append('Laser status: ' +
+            SapphireLaser.parseLaserStatus(self.send('?sta')))
+
+        for cmd, stat in [('?l', 'Ligh Emission on?'),
+                            ('?t', 'TEC Servo on?'),
+                            ('?k', 'Key Switch on?'),
+                            ('?sp', 'Target power:'),
+                            ('?p', 'Measured power:'),
+                            ('?hh', 'Head operating hours:')]:
+            result.append(stat + ' ' + self.send(cmd))
+
+        self._write('?fl')
+        faults = self._readline()
+        response = self._readline()
+        while response:
+            faults += ' ' + response
+            response = self._readline()
+
+        result.append(faults)
+        return result
+
+    @lock_comms
+    def _on_shutdown(self):
+        # Disable laser.
+        self._write('l=0')
+        self.flush_buffer()
+
+
+    ##  Initialization to do when cockpit connects.
+    @lock_comms
+    def initialize(self):
+        self.flush_buffer()
+
+
+    ## Turn the laser ON. Return True if we succeeded, False otherwise.
+    @lock_comms
+    def enable(self):
+        self._logger.info("Turning laser ON.")
+        # Turn on emission.
+        response = self._write('l=1')
+        self._logger.info("l=1: [%s]" % response)
+
+        # Enabling laser might take more than 500ms (default timeout)
+        prevTimeout = self.connection.timeout
+        self.connection.timeout = max(1, prevTimeout)
+        isON = self.get_is_on()
+        self.connection.timeout = prevTimeout
+
+        if not isON:
+            # Something went wrong.
+            self._logger.error("Failed to turn on. Current status:\r\n")
+            self._logger.error(self.get_status())
+            return False
+        return True
+
+
+    ## Turn the laser OFF.
+    @lock_comms
+    def disable(self):
+        self._logger.info("Turning laser OFF.")
+        return self._write('l=0')
+
+
+    ## Return True if the laser is currently able to produce light.
+    @lock_comms
+    def get_is_on(self, waitForResponse = 0):
+        return self.send('?l') == '1'
+
+
+    @lock_comms
+    def get_max_power_mw(self):
+        # '?maxlp' gets the maximum laser power in mW.
+        return float(self.send('?maxlp'))
+
+    @lock_comms
+    def get_min_power_mw(self):
+        # '?minlp' gets the minimum laser power in mW.
+        return float(self.send('?minlp'))
+
+    @lock_comms
+    def get_power_mw(self):
+        if not self.get_is_on():
+            return 0
+        return float(self.send('?p'))
```

Comment:
since this returns float, we should return `0.0` and not `0` if laser is not on.